<a href="https://colab.research.google.com/github/maryam98/ReDI-School/blob/main/Maryam_Jalali_Midterm_Test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


![titanic.jpg](https://image.geo.de/31780898/t/gt/v3/w960/r1.5/-/titanic---pa--122855442.jpg)
# Midterm Test 

Hello!

You already learned a lot how to explore and understand data. Let's test this! If you hand in your solution we will check it so you get direct feedback and know where you stand. To get feedback save a copy and rename it, so it contains your name. Then save the copy to this 
[folder](https://drive.google.com/drive/folders/1JqOTlRt9VhK3NVHt0svBHLI8DnMdpp5Q?usp=sharing).

The deadline for handing in your solution is May 9.

You can answer those questions as you wish by just coding or inspecting the data through a plot. But please answer the question with a sentence or two where you explain what the result is telling you. 

If you have any questions, please ask us on slack. We are happy to help.

Have fun coding!

### Titanic Dataset

Today we will explore a new dataset, the infamous Titanic dataset. It contains informations on passengers of the Titanic. At kaggle many people already investigated this dataset. Check it out to get new ideas: -> https://www.kaggle.com/c/titanic/overview


The dataset contains 12 columns.

- survival: 	Survival 	0 = No, 1 = Yes
- pclass: 	Ticket class 	1 = 1st, 2 = 2nd, 3 = 3rd
- sex: 	Sex 	
- Age: 	Age in years 	
- sibsp: 	# of siblings / spouses aboard the Titanic 	
- parch: 	# of parents / children aboard the Titanic 	
- ticket: 	Ticket number 	
- fare: 	Passenger fare 	
- cabin: 	Cabin number 	
- embarked: 	Port of Embarkation 	C = Cherbourg, Q = Queenstown, S = Southampton

It can be downloaded from: https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv


### 1. Load Libraries and data

Load libraries you might need and load the data using pandas. Get a first overview:
- Is the data loaded correctly?
- How does it look like?
- How many rows, how many columns?
- Do we see any Null values? What could they mean?

In [ ]:
!pip install dtale

In [ ]:
import pandas as pd
import seaborn as sns
import numpy as np
import dtale 
import dtale.app as dtale_app

In [ ]:
dtale.show(df)

http://12f6e6e86769:40000/dtale/main/1

In [ ]:
# Read csv into Pandas Dataframe and store in dataset variable
df = pd.read_csv("https://raw.githubusercontent.com/maryam98/ReDI-School/main/titanic.csv")

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


## The training-set has 891 rows and 12 columns . 2 of the features are floats, 5 are integers and 5 are objects

In [ ]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## In the train data, there’re 891 passengers, and the average survival rate is 38%. Age ranges from 0.42 to 80 and the average is ~30 year old. At least 50% of passengers don’t have siblings / spouses aboard the Titanic, and at least 75% of passengers don’t have parents / children aboard the Titanic. The fare varies a lot

In [ ]:
df.head(30)

PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
12           13         0       3   
13           14         0       3   
14           15         0       3   
15           16         1       2   
16           17         0       3   
17           18         1       2   
18           19         0       3   
19           20         1       3   
20           21         0       2   
21           22         1       2   
22           23         1       3   
23           24         1       1   
24           25         0       3   
25           26         1       3   
26           27         0       3   
27           28         0       1   
28           29         1       3   
29           30         0       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
5                                    Moran, Mr. James    male   NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male   NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female   NaN      0   
20                               Fynney, Mr. Joseph J    male  35.0      0   
21                              Beesley, Mr. Lawrence    male  34.0      0   
22                        McGowan, Miss. Anna "Annie"  female  15.0      0   
23                       Sloper, Mr. William Thompson    male  28.0      0   
24                      Palsson, Miss. Torborg Danira  female   8.0      3   
25  Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0      1   
26                            Emir, Mr. Farred Chehab    male   NaN      0   
27                     Fortune, Mr. Charles Alexander    male  19.0      3   
28                      O'Dwyer, Miss. Ellen "Nellie"  female   NaN      0   
29                                Todoroff, Mr. Lalio    male   NaN      0   

    Parch            Ticket      Fare        Cabin Embarked  
0       0         A/5 21171    7.2500          NaN        S  
1       0          PC 17599   71.2833          C85        C  
2       0  STON/O2. 3101282    7.9250          NaN        S  
3       0            113803   53.1000         C123        S  
4       0            373450    8.0500          NaN        S  
5       0            330877    8.4583          NaN        Q  
6  

In [ ]:
(df.isna().sum()/df.shape[0]).sort_values(ascending=False)

Cabin          0.771044
Age            0.198653
Embarked       0.002245
PassengerId    0.000000
Survived       0.000000
Pclass         0.000000
Name           0.000000
Sex            0.000000
SibSp          0.000000
Parch          0.000000
Ticket         0.000000
Fare           0.000000
dtype: float64

## ~20% of Age data is missing, 77% of Cabin data is missing, and 0.2% of Embarked data is missing

### 2. What is the average fare per class? On which range do the fares per class spread?

In [ ]:
df.groupby('Pclass')[['Fare']].describe()

Fare                                                              
        count       mean        std  min       25%      50%   75%       max
Pclass                                                                     
1       216.0  84.154687  78.380373  0.0  30.92395  60.2875  93.5  512.3292
2       184.0  20.662183  13.417399  0.0  13.00000  14.2500  26.0   73.5000
3       491.0  13.675550  11.778142  0.0   7.75000   8.0500  15.5   69.5500

##  the average fare per class1 is ~84, and is in the price range between 0 and 512
##the average fare per class2 is ~21, and is in the price range between 0 and ~74
 ## the average fare per class3 is ~14, and is in the price range between 0 and ~70

### 3. How many people had a cabin? 

In [ ]:
#solution 1:
df.groupby('Cabin')['PassengerId'].count().sum()

204

In [ ]:
#solution 2:
df.Cabin.notnull().value_counts()

False    687
True     204
Name: Cabin, dtype: int64

In [ ]:
#solution 3:
df.Cabin.count().sum()

204

In [ ]:
#solution 4:
df.Cabin.describe()

count         204
unique        147
top       B96 B98
freq            4
Name: Cabin, dtype: object

## in the Titanic 204 people had cabin

### 4. At which port lived the richest?

In [ ]:
df.groupby('Pclass')['Embarked'].describe()

count unique top freq
Pclass                      
1        214      3   S  127
2        184      3   S  164
3        491      3   S  353

In [ ]:
(df.loc[df.Pclass==1,'Embarked'].value_counts())*100/(df.loc[df.Pclass==1,'Embarked'].count())

S    59.345794
C    39.719626
Q     0.934579
Name: Embarked, dtype: float64

In [ ]:
df.loc[df.Pclass==1,'Embarked'].value_counts()

S    127
C     85
Q      2
Name: Embarked, dtype: int64

In [ ]:
(df.loc[df.Pclass==1,'Embarked'].value_counts())*100/(df.loc[:,'Embarked'].value_counts())

S    19.720497
C    50.595238
Q     2.597403
Name: Embarked, dtype: float64

**Here is  % of passengers in 1st class, among the passengers from each harbor.**

##  Cherbourg harbor  has the highest % of passengers in 1st class(~51%)

### 5. What is the most frequent last name?


In [ ]:
mylist=df.Name.str.split(',',expand=True)[[0]]
mylist.value_counts().to_frame('repetition')

repetition
0                        
Andersson               9
Sage                    7
Skoog                   6
Carter                  6
Goodwin                 6
...                   ...
Henry                   1
Hewlett                 1
Hirvonen                1
Hodges                  1
van Melkebeke           1

[667 rows x 1 columns]

##  "Andersson" is the most frequent last name Which has been repeated 9 times

### 6. Not for every person the age is noted. Fill in the gaps with the average age per gender.

In [ ]:
#solution1:
woman=df.loc[df.Sex=='female','Age'].mean()
man=df.loc[df.Sex=='male','Age'].mean()
df.update(df.loc[df.Sex=='male','Age'].fillna(man))
df.update(df.loc[df.Sex=='female','Age'].fillna(woman))

In [ ]:
#solution2:
df['Age'] = df['Age'].fillna(df.groupby('Sex')['Age'].transform('mean'))

In [ ]:
df.head(12)

PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   

                                                 Name     Sex        Age  \
0                             Braund, Mr. Owen Harris    male  22.000000   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.000000   
2                              Heikkinen, Miss. Laina  female  26.000000   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.000000   
4                            Allen, Mr. William Henry    male  35.000000   
5                                    Moran, Mr. James    male  30.726645   
6                             McCarthy, Mr. Timothy J    male  54.000000   
7                      Palsson, Master. Gosta Leonard    male   2.000000   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.000000   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.000000   
10                    Sandstrom, Miss. Marguerite Rut  female   4.000000   
11                           Bonnell, Miss. Elizabeth  female  58.000000   

    SibSp  Parch            Ticket     Fare Cabin Embarked  
0       1      0         A/5 21171   7.2500   NaN        S  
1       1      0          PC 17599  71.2833   C85        C  
2       0      0  STON/O2. 3101282   7.9250   NaN        S  
3       1      0            113803  53.1000  C123        S  
4       0      0            373450   8.0500   NaN        S  
5       0      0            330877   8.4583   NaN        Q  
6       0      0             17463  51.8625   E46        S  
7       3      1            349909  21.0750   NaN        S  
8       0      2            347742  11.1333   NaN        S  
9       1      0            237736  30.0708   NaN        C  
10      1      1           PP 9549  16.7000    G6        S  
11      0      0            113783  26.5500  C103        S

## the avrage of women's age is ~28 and the empty value of the age column for women is filled with this value
## the avrage of men's age is ~31 and the empty value of the age column for men is filled with this value

### Bonus: Upload this file as .ipnyb - notebook in your personal git-hub account and share the address here.

In [ ]:
# address: https://github.com/maryam98/ReDI-School/blob/main/Maryam_Jalali_Midterm_Test.ipynb